src/core/Core.js

```diff
@@ -35,12 +30,181 @@ export default class Core {
 
     this.translator = new Translator({locales: this.opts.locales})
     this.i18n = this.translator.translate.bind(this.translator)
-    // console.log(this.i18n('filesChosen', {smart_count: 3}))
 
     // Set up an event EventEmitter
     this.emitter = new ee.EventEmitter()
 
-    this.totalFilesSelectedCount = 0
+    this.defaultState = {
+      selectedFiles: {},
+      uploadedFiles: {},
+      modal: {
+        isHidden: true,
+        targets: []
+      }
+    }
+
+    this.state = Object.assign({}, this.state, this.defaultState)
+  }
+
+  /**
+   * Iterate on all plugins and run `update` on them. Called when state changes
+   *
+   */
+  updateAll () {
+    Object.keys(this.plugins).forEach((pluginType) => {
+      this.plugins[pluginType].forEach((plugin) => {
+        plugin.update(this.state)
+      })
+    })
+  }
+
+  /**
+   * Reset state to defaultState, used when Modal is closed, for example
+   *
+   */
+  resetState () {
+    this.setState(this.defaultState)
+  }
+
+  /**
+   * Updates state
+   *
+   * @param {newState} object
+   */
+  setState (newState) {
+    this.log(`Update state with: ${newState}`)
+    this.state = Object.assign({}, this.state, newState)
+    this.updateAll()
+  }
+
+  /**
+   * Gets current state, making sure to make a copy of the state object and pass that,
+   * instead of an actual reference to `this.state`
+   *
+   */
+  getState () {
+    return Object.assign({}, this.state)
```

Comment:
Looked through React docs:

> NEVER mutate this.state directly, as calling setState() afterwards may replace the mutation you made. Treat this.state as if it were immutable.
>
> setState() does not immediately mutate this.state but creates a pending state transition. Accessing this.state after calling this method can potentially return the existing value.
There is no guarantee of synchronous operation of calls to setState and calls may be batched for performance gains.
>
> setState() will always trigger a re-render unless conditional rendering logic is implemented in shouldComponentUpdate(). If mutable objects are being used and the logic cannot be implemented in shouldComponentUpdate(), calling setState() only when the new state differs from the previous state will avoid unnecessary re-renders.

https://facebook.github.io/react/docs/component-api.html